pkg/source/source.go

```diff
@@ -68,127 +77,184 @@ var _ Source = &Channel{}
 // (e.g. GitHub Webhook callback).  Channel requires the user to wire the external
 // source (eh.g. http handler) to write GenericEvents to the underlying channel.
 type Channel struct {
-	// once ensures the event distribution goroutine will be performed only once
-	once sync.Once
-
 	// Source is the source channel to fetch GenericEvents
 	Source <-chan event.GenericEvent
 
-	// stop is to end ongoing goroutine, and close the channels
-	stop <-chan struct{}
-
-	// dest is the destination channels of the added event handlers
-	dest []chan event.GenericEvent
-
 	// DestBufferSize is the specified buffer size of dest channels.
 	// Default to 1024 if not specified.
 	DestBufferSize int
 
-	// destLock is to ensure the destination channels are safely added/removed
-	destLock sync.Mutex
+	mu      sync.Mutex
+	stopped bool
+
+	// destinations are the destination channels of the added event handlers
+	destinations []chan event.GenericEvent
+
+	ctx    context.Context
+	cancel context.CancelFunc
+	group  *errgroup.Group
 }
 
 func (cs *Channel) String() string {
 	return fmt.Sprintf("channel source: %p", cs)
 }
 
-// Start implements Source and should only be called by the Controller.
-func (cs *Channel) Start(
-	ctx context.Context,
-	handler handler.EventHandler,
-	queue workqueue.RateLimitingInterface,
-	prct ...predicate.Predicate) error {
+// Start is internal and should be called only by the Controller to register an EventHandler with the Informer
+// to enqueue reconcile.Requests. It should NOT block, instead, it should start a goroutine and return immediately.
+// The context passed to Start can be used to cancel the blocking operations in the Start method. To cancel the
+// goroutine/ shutdown the source a user should call Stop.
+func (cs *Channel) Start(ctx context.Context, handler handler.EventHandler, queue workqueue.RateLimitingInterface,
```

Comment:
Nit: can we document for both sources how often / when start/stop can be called?

If I see correctly the difference is mainly that for source.Channel Start can be called multiple times to add multiple event handlers and on source.Kind Start can be called only once (and thus Kind allows only adding one event handler?)